README.md
[<p align="center"><img src="doc/logo.png" alt="node-apn" width="450" height="auto"></p>][node-apn]

node-apn
========

> A Node.js module for interfacing with the Apple Push Notification service.

<a href="https://github.com/parse-community/node-apn/actions?query=workflow%3Aci+branch%3Amaster">
  <img alt="Build status" src="https://github.com/parse-community/node-apn/workflows/ci/badge.svg?branch=master">
</a>

[logo]:doc/logo.png
[npm-image]:https://nodei.co/npm/@parse/node-apn.png?downloads=true
[npm-url]:https://npmjs.com/package/@parse/node-apn
[ci-image]:https://api.travis-ci.org/parse-community/node-apn.svg
[ci-url]:https://travis-ci.org/parse-community/node-apn

## Features

- Based on HTTP/2 based provider API
- Maintains a connection to the server to maximise notification batching and throughput.
- Automatically resends unsent notifications if an error occurs

## Installation

[npm][] is the preferred installation method:

```bash
$ npm install @parse/node-apn --save
```

## Quick Start

This readme is a brief introduction, please refer to the full [documentation](doc/apn.markdown) in `doc/` for more details.

If you have previously used v1.x and wish to learn more about what's changed in v2.0, please see [What's New](doc/whats-new.markdown)

### Load in the module

```javascript
var apn = require('@parse/node-apn');
```

### Connecting
Create a new connection to the Apple Push Notification provider API, passing a dictionary of options to the constructor. You must supply your token credentials in the options.

```javascript
var options = {
  token: {
    key: "path/to/APNsAuthKey_XXXXXXXXXX.p8",
    keyId: "key-id",
    teamId: "developer-team-id"
  },
  production: false
};

var apnProvider = new apn.Provider(options);
```

By default, the provider will connect to the sandbox unless the environment variable `NODE_ENV=production` is set.

For more information about configuration options consult the [provider documentation](doc/provider.markdown).

Help with preparing the key and certificate files for connection can be found in the [wiki][certificateWiki]

#### Connecting through an HTTP proxy

If you need to connect through an HTTP proxy, you simply need to provide the `proxy: {host, port}` option when creating the provider. For example:

```javascript
var options = {
  token: {
    key: "path/to/APNsAuthKey_XXXXXXXXXX.p8",
    keyId: "key-id",
    teamId: "developer-team-id"
  },
  proxy: {
    host: "192.168.10.92",
    port: 8080
  }
  production: false
};

var apnProvider = new apn.Provider(options);
```

The provider will first send an HTTP CONNECT request to the specified proxy in order to establish an HTTP tunnel. Once established, it will create a new secure connection to the Apple Push Notification provider API through the tunnel.

### Sending a notification
To send a notification you will first need a device token from your app as a string

```javascript
let deviceToken = "a9d0ed10e9cfd022a61cb08753f49c5a0b0dfb383697bf9f9d750a1003da19c7"
```

Create a notification object, configuring it with the relevant parameters (See the [notification documentation](doc/notification.markdown) for more details.)

```javascript
var note = new apn.Notification();

note.expiry = Math.floor(Date.now() / 1000) + 3600; // Expires 1 hour from now.
note.badge = 3;
note.sound = "ping.aiff";
note.alert = "\uD83D\uDCE7 \u2709 You have a new message";
note.payload = {'messageFrom': 'John Appleseed'};
note.topic = "<your-app-bundle-id>";
```

Send the notification to the API with `send`, which returns a promise.

```javascript
apnProvider.send(note, deviceToken).then( (result) => {
  // see documentation for an explanation of result
});
```

This will result in the the following notification payload being sent to the device

```json
{"messageFrom":"John Appelseed","aps":{"badge":3,"sound":"ping.aiff","alert":"\uD83D\uDCE7 \u2709 You have a new message"}}
```

You should only create one `Provider` per-process for each certificate/key pair you have. You do not need to create a new `Provider` for each notification. If you are only sending notifications to one app then there is no need for more than one `Provider`.

If you are constantly creating `Provider` instances in your app, make sure to call `Provider.shutdown()` when you are done with each provider to release its resources and memory.

## Troubleshooting

You are encouraged to read the extremely informative [Troubleshooting Push Notifications][tn2265] Tech Note in the first instance, in case your query is answered there.

## Support

If you have any questions or difficulties working with the module, the [node-apn Google group][googlegroup] should be your first port of call.

Please include as much detail as possible - especially debug logs. If the problem is reproducible, sample code is also extremely helpful. GitHub Issues should only be created for verified problems and enhancements, this will allow them to be tracked more easily.

## Resources

* [Local and Push Notification Programming Guide: APNs Overview][pl]
* [Apple Technical Note: Troubleshooting Push Notifications][tn2265]
* [List of Projects, Applications and Companies Using Node-apn][pacapn]

## Credits

Created by [Andrew Naylor][argon]

Thanks to: [Ian Babrou][bobrik], [dgthistle][dgthistle], [Keith Larsen][keithnlarsen], [Mike P][mypark], [Greg Bergé][neoziro], [Asad ur Rehman][AsadR], [Nebojsa Sabovic][nsabovic], [Alberto Gimeno][gimenete], [Randall Tombaugh][rwtombaugh], [Michael Stewart][thegreatmichael], [Olivier Louvignes][mgcrea], [porsager][porsager], [Craig Hockenberry][chockenberry]

## License

Released under the MIT License

> Copyright (c) 2013 Andrew Naylor
>
> Permission is hereby granted, free of charge, to any person obtaining a copy
> of this software and associated documentation files (the "Software"), to deal
> in the Software without restriction, including without limitation the rights
> to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
> copies of the Software, and to permit persons to whom the Software is
> furnished to do so, subject to the following conditions:
>
> The above copyright notice and this permission notice shall be included in
> all copies or substantial portions of the Software.
>
> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

[npm]: https://npmjs.org
[node-apn]: https://github.com/parse-community/node-apn
[certificateWiki]:https://github.com/node-apn/node-apn/wiki/Preparing-Certificates "Preparing Certificates"
[pl]: https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html "Local and Push Notification Programming Guide: APNs Overview"
[tn2265]: http://developer.apple.com/library/ios/#technotes/tn2265/_index.html "Troubleshooting Push Notifications"
[googlegroup]:https://groups.google.com/group/node-apn "node-apn Google Group"
[pacapn]:https://github.com/node-apn/node-apn/wiki/Projects,-Applications,-and-Companies-Using-Node-apn "List of Projects, Applications and Companies Using node-apn"
[argon]: https://github.com/argon
[bobrik]: https://github.com/bobrik
[dgthistle]: https://github.com/dgthistle
[keithnlarsen]: https://github.com/keithnlarsen
[mypark]: https://github.com/mypark
[neoziro]: https://github.com/neoziro
[AsadR]: https://github.com/AsadR
[nsabovic]: https://github.com/nsabovic
[gimenete]: https://github.com/gimenete
[rwtombaugh]: https://github.com/rwtombaugh
[thegreatmichael]: https://github.com/thegreatmichael
[mgcrea]: https://github.com/mgcrea
[porsager]: https://github.com/porsager
[chockenberry]: https://github.com/chockenberry



ChangeLog.md
## Changelog

4.1.1
* FIX: add proxy and fix logger types

4.1.0
* NEW: Proxy support
* FIX: Close client on unexpected frame error
* NEW: Add a configurable info/error logger
* FIX: Update endpoint address
* FIX: Add support to zero valued expiry

4.0.0
* BREAKING: shutdown() function on client now destroy the client and future requests throw error
* Add test cases to Client
* Improve verbose messages
* Add request timeout handler
* Add a new config setting requestTimeout (defaults to 5000 milliseconds)
* Change ResponseFailure.status typing to be number
* Fix nits on typos in documentation
* Also run tests in Node 12 and 14
* Remove the unreferenced lib/protocol directory
* Improve GO AWAY handler

3.2.0
* Update node-forge and pin packages
* Add typings for the "voip" NotificationPushType
* Typed Notification.pushType

3.1.0
* Fix: http2session.shutdown() is not a function
* Update npm install command in README
* Add support for handling http2 GOAWAY messages and set http2 ping interval

3.0.6
* Set npm access to public in package.json to override default of private for scoped packages

3.0.5
* Change Travis to use node 10 instead of node 9
* Attempt changes to travis to get it to publish to npm

3.0.4
* No change, bumping version with travis enabled

3.0.3
* Create an npm installable package `@parse/node-apn`

3.0.2-parse:
* Fix 3.0.1-parse release

3.0.1-parse:
* Add support for apns-push-type key

3.0.0:
* Native Node.js HTTP/2 client
* Requires Node.js 8.9.1 or later
* Breaking changes:
    * The `status` property of notification results has changed from `string` to `number`

2.1.5:
* Improve typescript type definitions (#536)
* Several documentation improvements (#537, #547, #550, #555, #560, #560)
* Update dependencies (#566)
* Update .gitignore and add .nvmrc (#573)

2.1.4:
* Don't mutate passed in payload object
* Change APNs development endpoint to api.development.push.apple.com
* Update node-http2 fork

2.1.3:
* Add thread-id to notification payload (#484)
* Workaround an issue where APNS would return an HTTP 500 and make the connection unusable (#449, #480)
* Regularly renew JSON web token (#449, #480)

2.1.2:
* Add TypeScript type definitions (#441, #465)
* Update node-http2 fork to 3.3.6 and add stream.upstream not defined check (#444)

2.1.1:
* Fix broken logger (#438)
* Use archive URL for forked node-http2 module (#433)

2.1.0:

* JSON Web Token support (#401)
* Add `threadId` property for notifications
* Fix a leak caused when an error is emitted from the endpoint manager
    * Queued notifications would resolve as errored but would remain in the queue, leading them to be repeatedly rejected without termination and causing an error log from node

2.0.0:

* HTTP/2 Provider API (#277)
* `apn.Connection` has been renamed to `apn.Provider`
* `apn.Feedback` has been removed
* `apn.Device` has been removed - all tokens are now hex-encoded strings
* `apn.token` is provided to validate tokens and convert from `Buffer` if
  necessary
* Notifications are now required to have an associated `topic`
* `pushNotification(notification, tokens)` is now simply, `send(notification, recipients)`
* `send` returns a promise which will be fulfilled when all notifications have
  been sent (#362)
* Reliable Delivery (#168)
* Ability to mock `apn.Provider` behaviour
* Notification can be created in a "finished" state (#240)

1.7.8:

* Add support for Mutable Content (#413)

1.7.7:

* Fix Documentation URL (#393)
* Prevent `badge: 0` sending `badge: undefined` (#394)
* Avoid error, when deferredConnection is null (#397)

1.7.6:

* Emit an error when retry limit is exceeded (#333)
* Documentation fixes (#229, #343, #379)
* Reinstate broken aps behaviour (#377)

1.7.5:

* Notification property improvements
* Round-robin all open sockets when sending notifications
* `minConnections` option
* Prevent crashing when cancelling feedback multiple times (#307)
* Fix connection.shutdown behaviour (#299)
* Small doc improvements

1.7.4:

* Notification trim fixes. (#220)
* Notification refactoring
* `debug` package is now a dependency (#267)

1.7.3:

* Added support for increased payload length for VoIP applications (Closes #207)
* Fixed a bug with trimming UTF-16 encoded payloads
* Dropped support for node v0.6 as it doesn't support UTF-16 surrogate pairs. Plus it's old. It'll still work if needed though, if you use UTF-8.

1.7.2:

* Fixed: #238, only emit `error` when the problem is unrecoverable. Any use of `node-apn` should have an `error` listener attached to prevent uncaught exceptions.
* Various coding style improvements.
* CI.
* Removed legacy protocol support. I don't see any reason to keep it around, let me know if this causes you any problems.

1.7.1:

* Fixed: #224, always passing a CA value even if no certificates were specified. In this case the TLS library will not use the built in root certificates and will always fail to trust the server certificate.
* Changed: Socket timeout default has changed for disabled to 1 hour.
* Documentation fixes

1.7.0:

* Added: Credential validator to catch common configuration errors.
* Fixed: Documentation errors.

1.6.2:

* Updated maximum payload size to 2048 bytes. (See #181).
* Fixed: Feedback wouldn't include error message when emitting `feedbackError`.

1.6.1:

* Increased default cache length to 1000; Previous value of 100 was overly cautious.
* Added a delay of 100ms to connection to ensure notifications aren't lost if wrong certificates are used.
* Fixes #195: Better handling of socket creation
* Fixes #200: Improved buffer handling code when multiple connections are enabled.
* Minor optimisation to Notification processing, removing the need for 2 `JSON.stringify` calls
* Fixes #196: Check cache has contents before trying to access. Also ensure minimum size is "1" to allow "transmitted" events to be emitted.
* Fixes #199: Emit a "drain" event when no further notifications require sending. Most useful in a batch environment.

1.6.0:


test/provider.js
"use strict";

const sinon = require("sinon");
const EventEmitter = require("events");

describe("Provider", function() {
  let fakes, Provider;

  beforeEach(function () {
    fakes = {
      Client: sinon.stub(),
      client: new EventEmitter(),
    };

    fakes.Client.returns(fakes.client);
    fakes.client.write = sinon.stub();
    fakes.client.shutdown = sinon.stub();

    Provider = require("../lib/provider")(fakes);
  });

  describe("constructor", function () {

    context("called without `new`", function () {
      it("returns a new instance", function () {
        expect(Provider()).to.be.an.instanceof(Provider);
      });
    });

    describe("Client instance", function() {
      it("is created", function () {
        Provider();

        expect(fakes.Client).to.be.calledOnce;
        expect(fakes.Client).to.be.calledWithNew;
      });

      it("is passed the options", function () {
        const options = { "configKey": "configValue"};

        Provider(options);
        expect(fakes.Client).to.be.calledWith(options);
      });
    });
  });

  describe("send", function () {

    describe("single notification behaviour", function () {
      let provider;

      context("transmission succeeds", function () {
        beforeEach( function () {
          provider = new Provider( { address: "testapi" } );

          fakes.client.write.onCall(0).returns(Promise.resolve({ device: "abcd1234" }));
        });

        it("invokes the writer with correct `this`", function () {
          return provider.send(notificationDouble(), "abcd1234")
            .then(function () {
              expect(fakes.client.write).to.be.calledOn(fakes.client);
            });
        });

        it("writes the notification to the client once", function () {
          return provider.send(notificationDouble(), "abcd1234")
            .then(function () {
              const notification = notificationDouble();
              const builtNotification = {
                headers: notification.headers(),
                body: notification.compile(),
              };
              expect(fakes.client.write).to.be.calledOnce;
              expect(fakes.client.write).to.be.calledWith(builtNotification, "abcd1234");
            });
        });

        it("does not pass the array index to writer", function () {
          return provider.send(notificationDouble(), "abcd1234")
            .then(function () {
              expect(fakes.client.write.firstCall.args[2]).to.be.undefined;
            });
        });

        it("resolves with the device token in the sent array", function () {
          return expect(provider.send(notificationDouble(), "abcd1234"))
            .to.become({ sent: [{"device": "abcd1234"}], failed: []});
        });
      });

      context("error occurs", function () {
        let promise;

        beforeEach(function () {
          const provider = new Provider( { address: "testapi" } );

          fakes.client.write.onCall(0).returns(Promise.resolve({ device: "abcd1234", status: "400", response: { reason: "BadDeviceToken" }}));
          promise = provider.send(notificationDouble(), "abcd1234");
        });

        it("resolves with the device token, status code and response in the failed array", function () {
          return expect(promise).to.eventually.deep.equal({ sent: [], failed: [{"device": "abcd1234", "status": "400", "response": { "reason" : "BadDeviceToken" }}]});
        });
      });
    });

    context("when multiple tokens are passed", function () {
      const connectionFailedError = new Error("connection failed");

      beforeEach(function () {
          fakes.resolutions = [
            { device: "abcd1234" },
            { device: "adfe5969", status: "400", response: { reason: "MissingTopic" }},
            { device: "abcd1335", status: "410", response: { reason: "BadDeviceToken", timestamp: 123456789 }},
            { device: "bcfe4433" },
            { device: "aabbc788", status: "413", response: { reason: "PayloadTooLarge" }},
            { device: "fbcde238", error: connectionFailedError },
          ];
      });

      context("streams are always returned", function () {
        let promise;

        beforeEach( function () {
          const provider = new Provider( { address: "testapi" } );

          for(let i=0; i < fakes.resolutions.length; i++) {
            fakes.client.write.onCall(i).returns(Promise.resolve(fakes.resolutions[i])); 
          }

          promise = provider.send(notificationDouble(), fakes.resolutions.map( res => res.device ));

          return promise;
        });

        it("resolves with the sent notifications", function () {
          return promise.then( (response) => {
            expect(response.sent).to.deep.equal([{device: "abcd1234"}, {device: "bcfe4433"}]);
          });
        });

        it("resolves with the device token, status code and response or error of the unsent notifications", function () {
          return promise.then( (response) => {
            expect(response.failed).to.deep.equal([
              { device: "adfe5969", status: "400", response: { reason: "MissingTopic" }},
              { device: "abcd1335", status: "410", response: { reason: "BadDeviceToken", timestamp: 123456789 }},
              { device: "aabbc788", status: "413", response: { reason: "PayloadTooLarge" }},
              { device: "fbcde238", error: connectionFailedError },
            ]);
          });
        });
      });
    });
  });

  describe("shutdown", function () {
    it("invokes shutdown on the client", function () { 
      let provider = new Provider({});
      provider.shutdown();

      expect(fakes.client.shutdown).to.be.calledOnce;
    });
  });
});

function notificationDouble() {
  return {
    headers: sinon.stub().returns({}),
    payload: { aps: { badge: 1 } },
    compile: function() { return JSON.stringify(this.payload); }
  };
}


index.d.ts
/// <reference types="node" />

import { EventEmitter } from 'events';

export interface ProviderToken {
  /**
   * The filename of the provider token key (as supplied by Apple) to load from disk, or a Buffer/String containing the key data.
   */
  key: Buffer|string;
  /**
   * The ID of the key issued by Apple
   */
  keyId: string;
  /**
   * ID of the team associated with the provider token key
   */
  teamId: string;
}

export interface ProviderOptions {
  /**
   * Configuration for Provider Authentication Tokens. (Defaults to: null i.e. fallback to Certificates)
   */
  token?: ProviderToken;
  /**
   * The filename of the connection certificate to load from disk, or a Buffer/String containing the certificate data. (Defaults to: `cert.pem`)
   */
  cert?: string|Buffer;
  /**
   * The filename of the connection key to load from disk, or a Buffer/String containing the key data. (Defaults to: `key.pem`)
   */
  key?: string|Buffer;
  /**
   * An array of trusted certificates. Each element should contain either a filename to load, or a Buffer/String (in PEM format) to be used directly. If this is omitted several well known "root" CAs will be used. - You may need to use this as some environments don't include the CA used by Apple (entrust_2048).
   */
  ca?: (string|Buffer)[];
  /**
   * File path for private key, certificate and CA certs in PFX or PKCS12 format, or a Buffer containing the PFX data. If supplied will always be used instead of certificate and key above.
   */
  pfx?: string|Buffer;
  /**
   * The passphrase for the connection key, if required
   */
  passphrase?: string;
  /**
   * Specifies which environment to connect to: Production (if true) or Sandbox - The hostname will be set automatically. (Defaults to NODE_ENV == "production", i.e. false unless the NODE_ENV environment variable is set accordingly)
   */
  production?: boolean;
  /**
   * Reject Unauthorized property to be passed through to tls.connect() (Defaults to `true`)
   */
  rejectUnauthorized?: boolean;
  /**
   * The maximum number of connection failures that will be tolerated before `apn` will "terminate". (Defaults to: 3)
   */
  connectionRetryLimit?: number;
  /**
   * The maximum time in ms that apn will wait for a request to be responded. (Defaults to: 5000)
   */
  requestTimeout?: number;
}

interface ApsAlert {
  body?: string
  "loc-key"?: string
  "loc-args"?: any[]
  title?: string
  "title-loc-key"?: string
  "title-loc-args"?: any[]
  action?: string
  "action-loc-key"?: string
}

interface Aps {
  alert?: string | ApsAlert
  "launch-image"?: string
  badge?: number
  sound?: string
  "content-available"?: undefined | 1
  "mutable-content"?: undefined | 1
  "url-args"?: string[]
  category?: string
}

export interface ResponseSent {
  device: string;
}
export interface ResponseFailure {
  device: string;
  error?: Error;
  status?: number;
  response?: {
    reason: string;
    timestamp?: string;
  };
}

export interface Responses {
  sent:   ResponseSent[];
  failed: ResponseFailure[];
}

export class Provider extends EventEmitter {
  constructor(options: ProviderOptions);
  /**
   * This is main interface for sending notifications. Create a Notification object and pass it in, along with a single recipient or an array of them and node-apn will take care of the rest, delivering a copy of the notification to each recipient.
   *
   * A "recipient" is a String containing the hex-encoded device token.
   */
  send(notification: Notification, recipients: string|string[]): Promise<Responses>;
  /**
   * Indicate to node-apn that it should close all open connections when the queue of pending notifications is fully drained. This will allow your application to terminate.
   */
  shutdown(): void;
}

export type NotificationPushType = 'background' | 'alert' | 'voip';

export interface NotificationAlertOptions {
  title?: string;
  subtitle?: string;
  body: string;
  "title-loc-key"?: string;
  "title-loc-args"?: string[];
  "action-loc-key"?: string;
  "loc-key"?: string;
  "loc-args"?: string[];
  "launch-image"?: string;
}

export class Notification {
  /**
   * You can optionally pass in an object representing the payload, or configure properties on the returned object.
   */
  constructor(payload?: any);

  /**
   * Required: The destination topic for the notification.
   */
  public topic: string;
  /**
   * A UUID to identify the notification with APNS. If an id is not supplied, APNS will generate one automatically. If an error occurs the response will contain the id. This property populates the apns-id header.
   */
  public id: string;
  /**
   * The UNIX timestamp representing when the notification should expire. This does not contribute to the 2048 byte payload size limit. An expiry of 0 indicates that the notification expires immediately.
   */
  public expiry: number;
  /**
   * Provide one of the following values:
   *
   * - 10 - The push message is sent immediately. (Default)
   *   > The push notification must trigger an alert, sound, or badge on the device. It is an error use this priority for a push that contains only the content-available key.
   * - 5 - The push message is sent at a time that conserves power on the device receiving it.
   */
  public priority: number;

  public collapseId: string;
  public pushType: NotificationPushType;
  public threadId: string;

  /**
   * This Object is JSON encoded and sent as the notification payload. When properties have been set on notification.aps (either directly or with convenience setters) these are added to the payload just before it is sent. If payload already contains an aps property it is replaced.
   */
  public payload: any;
  public aps: Aps;

  /**
   * If supplied this payload will be encoded and transmitted as-is. The convenience setters will have no effect on the JSON output.
   */
  public rawPayload: any;

  /**
   * The value to specify for `payload.aps.badge`
   */
  public badge: number;
  /**
   * The value to specify for `payload.aps.sound`
   */
  public sound: string;
  /**
   * The value to specify for `payload.aps.alert` can be either a `String` or an `Object` as outlined by the payload documentation.
   */
  public alert: string|NotificationAlertOptions;
  /**
   * Setting this to true will specify "content-available" in the payload when it is compiled.
   */
  public contentAvailable: boolean;
  /**
   *
   */
  public mutableContent: boolean;
  /**
   * The value to specify for the `mdm` field where applicable.
   */
  public mdm: string|Object;
  /**
   * The value to specify for `payload.aps['url-args']`. This used for Safari Push NOtifications and should be an array of values in accordance with the Web Payload Documentation.
   */
  public urlArgs: string[];
}

export function token(token: (string | Buffer)) : string


lib/client.js
"use strict";

const VError = require("verror");
const extend = require("./util/extend");

module.exports = function (dependencies) {
  const logger = dependencies.logger;
  const config = dependencies.config;
  const http2 = dependencies.http2;

  const {
    HTTP2_HEADER_STATUS,
    HTTP2_HEADER_SCHEME,
    HTTP2_HEADER_METHOD,
    HTTP2_HEADER_AUTHORITY,
    HTTP2_HEADER_PATH,
    HTTP2_METHOD_POST,
    NGHTTP2_CANCEL,
  } = http2.constants;

  function Client (options) {
    this.config = config(options);
    this.healthCheckInterval = setInterval(() => {
      if (this.session && !this.session.closed && !this.session.destroyed && !this.isDestroyed) {
        this.session.ping((error, duration) => {
          if (error) {
            logger("No Ping response after " + duration + " ms with error:" + error.message);
            return;
          }
          logger("Ping response after " + duration + " ms");
        });
      }
    }, this.config.heartBeat).unref();
  }

  // Session should be passed except when destroying the client
  Client.prototype.destroySession = function (session, callback) {
    if (!session) {
      session = this.session;
    }
    if (session) {
      if (this.session === session) {
        this.session = null;
      }
      if (!session.destroyed) {
        session.destroy();
      }
    }
    if (callback) {
      callback();
    }
  }

  // Session should be passed except when destroying the client
  Client.prototype.closeAndDestroySession = function (session, callback) {
    if (!session) {
      session = this.session;
    }
    if (session) {
      if (this.session === session) {
        this.session = null;
      }
      if (!session.closed) {
        session.close(() => this.destroySession(session, callback));
      } else {
        this.destroySession(session, callback)
      }
    } else if (callback) {
      callback();
    }
  }

  Client.prototype.write = function write (notification, device, count) {
    if (this.isDestroyed) {
      return Promise.resolve({ device, error: new VError("client is destroyed") });
    }

    // Connect session
    if (!this.session || this.session.closed || this.session.destroyed) {
      const session = this.session = http2.connect(this._mockOverrideUrl || `https://${this.config.address}`, this.config);

      this.session.on("close", () => {
        if (logger.enabled) {
          logger("Session closed");
        }
        this.destroySession(session);
      });

      this.session.on("socketError", (error) => {
        if (logger.enabled) {
          logger(`Socket error: ${error}`);
        }
        this.closeAndDestroySession(session);
      });

      this.session.on("error", (error) => {
        if (logger.enabled) {
          logger(`Session error: ${error}`);
        }
        this.closeAndDestroySession(session);
      });

      this.session.on("goaway", (errorCode, lastStreamId, opaqueData) => {
        if (logger.enabled) {
          logger(`GOAWAY received: (errorCode ${errorCode}, lastStreamId: ${lastStreamId}, opaqueData: ${opaqueData})`);
        }
        this.closeAndDestroySession(session);
      });

      if (logger.enabled) {
        this.session.on("connect", () => {
          logger("Session connected");
        });

        this.session.on("frameError", (frameType, errorCode, streamId) => {
          logger(`Frame error: (frameType: ${frameType}, errorCode ${errorCode}, streamId: ${streamId})`);
        });
      }
    }

    let tokenGeneration = null;
    let status = null;
    let responseData = "";
    let retryCount = count || 0;

    const headers = extend({
      [HTTP2_HEADER_SCHEME]: "https",
      [HTTP2_HEADER_METHOD]: HTTP2_METHOD_POST,
      [HTTP2_HEADER_AUTHORITY]: this.config.address,
      [HTTP2_HEADER_PATH]: `/3/device/${device}`,
    }, notification.headers);

    if (this.config.token) {
      if (this.config.token.isExpired(3300)) {
        this.config.token.regenerate(this.config.token.generation);
      }
      headers.authorization = `bearer ${this.config.token.current}`;
      tokenGeneration = this.config.token.generation;
    }

    const request = this.session.request(headers)

    request.setEncoding("utf8");

    request.on("response", (headers) => {
      status = headers[HTTP2_HEADER_STATUS];
    });

    request.on("data", (data) => {
      responseData += data;
    });

    request.write(notification.body);

    return new Promise ( resolve => {
      request.on("end", () => {
        try {
          if (logger.enabled) {
            logger(`Request ended with status ${status} and responseData: ${responseData}`);
          }

          if (status === 200) {
            resolve({ device });
          } else if (responseData !== "") {
            const response = JSON.parse(responseData);

            if (status === 403 && response.reason === "ExpiredProviderToken" && retryCount < 2) {
              this.config.token.regenerate(tokenGeneration);
              resolve(this.write(notification, device, retryCount + 1));
              return;
            } else if (status === 500 && response.reason === "InternalServerError") {
              this.closeAndDestroySession();
              let error = new VError("Error 500, stream ended unexpectedly");
              resolve({ device, error });
              return;
            }

            resolve({ device, status, response });
          } else {
            let error = new VError("stream ended unexpectedly");
            resolve({ device, error });
          }
        } catch (e) {
          const error = new VError(e, 'Unexpected error processing APNs response');
          if (logger.enabled) {
            logger(`Unexpected error processing APNs response: ${e.message}`);
          }
          resolve({ device, error });
        }
      });

      request.setTimeout(this.config.requestTimeout, () => {
        if (logger.enabled) {
          logger('Request timeout');
        }

        request.close(NGHTTP2_CANCEL);

        resolve({ device, error: new VError("apn write timeout") });
      });

      request.on("aborted", () => {
        if (logger.enabled) {
          logger('Request aborted');
        }

        resolve({ device, error: new VError("apn write aborted") });
      });

      request.on("error", (error) => {
        if (logger.enabled) {
          logger(`Request error: ${error}`);
        }

        if (typeof error === "string") {
          error = new VError("apn write failed: %s", error);
        } else {
          error = new VError(error, "apn write failed");
        }
        resolve({ device, error });
      });

      if (logger.enabled) {
        request.on("frameError", (frameType, errorCode, streamId) => {
          logger(`Request frame error: (frameType: ${frameType}, errorCode ${errorCode}, streamId: ${streamId})`);
        });
      }

      request.end();
    });
  };

  Client.prototype.shutdown = function shutdown(callback) {
    if (this.isDestroyed) {
      if (callback) {
        callback();
      }
      return;
    }
    if (logger.enabled) {
      logger('Called client.shutdown()');
    }
    this.isDestroyed = true;
    if (this.healthCheckInterval) {
      clearInterval(this.healthCheckInterval);
      this.healthCheckInterval = null;
    }
    this.closeAndDestroySession(undefined, callback);
  };

  return Client;
}
